staging/src/k8s.io/apiserver/pkg/server/filters/routine.go

```diff
@@ -0,0 +1,78 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"context"
+	"net/http"
+
+	"k8s.io/apiserver/pkg/endpoints/request"
+	"k8s.io/klog/v2"
+)
+
+type taskKeyType int
+
+const taskKey taskKeyType = iota
+
+type Task struct {
+	F func()
+}
+
+func WithTask(parent context.Context, t *Task) context.Context {
+	return request.WithValue(parent, taskKey, t)
+}
+
+func TaskFrom(ctx context.Context) *Task {
+	t, _ := ctx.Value(taskKey).(*Task)
+	return t
+}
+
+// WithRoutine returns an http.Handler that executes preparation of long running requests (i.e. watches)
+// in a separate Goroutine and then serves the long running request in the main Goroutine. Doing so allows
+// freeing stack memory used in preparation Goroutine for better memory efficiency.
+func WithRoutine(handler http.Handler, longRunning request.LongRunningRequestCheck, resolver request.RequestInfoResolver) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		requestInfo, err := resolver.NewRequestInfo(req)
```

Comment:
The reason I created RequestInfo was to put this handler wrap at as top level as possible to have clean stack (ie. haven't gone into too many calls) so that when we go back to this handler, the stack memory usage is low, which is the idea for the optimization

Reply:
I'm not a fan of it - we shouldn't be duplicating the work and optimizing one thing at the cost of breaking something else that was optimized before.
This is also somewhat related to my comment about putting this handler in front of WithPanic...

Reply:
Okay, I moved the routine up in the chain to reuse the work. FTR, I observed difference of 2MiB in a quick test comparison for stack memory usage with 5k watch connections, which is completely acceptable given this PR will reduce from ~200MiB to ~60MiB.